### ISSUE TITLE
Adding in AlphaZero changes

### ISSUE THREAD

**Issue by justinwong098** [gcp/leela-zero#267]
DeepMind recently released a paper for their new program, AlphaZero. One of the major changes was "omitting the evaluation step and the selection of best player". If I'm interpreting this correctly, this means that they use the latest network to generate new games, even if it isn't the strongest so far. Is this a reasonable modification to add to LeelaZero?

Here's the quote from the paper:

> In AlphaGo Zero, self-play games were generated by the best player from all previous iterations. After each iteration of training, the performance of the new player was measured against the best player; if it won by a margin of 55% then it replaced the best player and self-play games were subsequently generated by this new player. In contrast, AlphaZero simply maintains a single neural network that is updated continually, rather than waiting for an iteration to complete. Self-play games are generated by using the latest parameters for this neural network, omitting the evaluation step and the selection of best player.

**Comment by kityanhem** [comment]
https://arxiv.org/abs/1712.01815
the document about AlphaZero

**Comment by isty2e** [comment]
For two reasons, I don't think this applicable to LZ.

1. The balance between speeds of training and self-play seems important fo me. If training is slower, the quality of network will not increase sufficiently fast, thereby resulting in self-plays of lower quality. On the other hand, if self-play generation is slower, then the network is likely to be overfitted, and considering that less training can be more desirable, the result can be questionable.

2. I don't think given the number of self-play games AlphaZero is stronger with AlphaGo Zero. They have generated 21M games for go, whereas it was 5M for AlphaGo Zero compared in the paper. AlphaZero reached AlphaGo Zero level after 10M games or so.

3. It will be much harder to streamline the process, deploying the NN which changes over time continuously.

**Comment by Ttl** [comment]
This paper (https://arxiv.org/abs/1705.08439), which uses basically the same algorithm as AGZ for Hex doesn't use evaluation either. However they say that the reason for it is because they found that the new net was consistently stronger and they didn't find that evaluation was necessary. The same can't be said about LZ right now.

**Comment by justinwong098** [comment]
>I don't think given the number of self-play games AlphaZero is stronger with AlphaGo Zero.

You're right. I didn't think about this. You're probably right that this AlphaZero change shouldn't be implemented. However, to be fair, AlphaZero also had a few other changes that made it less capable of playing Go: such as the fact they didn't enrich it with 8-fold symmetry, or the fact that AlphaGo Zero had tuned hyperparameters, while AlphaZero didn't, or that AlphaZero had to take into account other outcomes such as draws. I'd imagine that including 8 symmetries would greatly enhance the training data available, which AlphaZero didn't take advantage of.

**Comment by isty2e** [comment]
@justinwong098 It is a bit ambiguously written, but I understood it as the symmetry was exploited in AlphaZero for go, but not for chess and shogi. At any rate, I personally consider the first reason the most important one here.

**Comment by justinwong098** [comment]
@isty2e 
I would consider exploiting symmetries to be a part of Domain Knowledge, of which symmetry isn't mentioned in that section. 
I'm not sure that training and self-play are matching very well right now. It seems GCP is spending a lot of time evaluating. But I could be wrong.

If we were to implement these AlphaZero changes, my (uneducated) suggestion would be to continue training new networks every 25-30k games, as we've been doing. But instead of evaluating whether the new network is stronger, just use the new network to generate new self-play games.

**Comment by gcp** [comment]
Dropping evaluation would make things a lot easier, as it's a major source of latency. There's work being done to distribute that, but still.

It's surprising to me this approach works. We've seen the playing strength drop greatly during training.

Also they used 800 MCTS simulations. So 1600 vs 1000 and all that is clearly not relevant.

**Comment by isty2e** [comment]
@justinwong098 It is written like this:

>The rules of Go are invariant to rotation and reﬂection. This fact was exploited in AlphaGo and AlphaGo Zero in two ways. First, training data was augmented by generating 8 symmetries for each position. Second, during MCTS, board positions were transformed using a randomly selected rotation or reﬂection before being evaluated by the neural network, so that the Monte-Carlo evaluation is averaged over different biases. The rules of chess and shogi are asymmetric, and in general symmetries cannot be assumed. AlphaZero does not augment the training data and does not transform the board position during MCTS.

So my understanding is that, AlphaZero takes advantage of symmetry ("This fact was exploited in AlphaGo and AlphaGo Zero in two ways."), but it is not the case for chess and shogi.

Concerning the balancing between two processes, the unbalance could be more detrimental to AlphaZero than AlphaGo. Of course it is just my impression, but will be very likely. Also I feel dubious if such a continuous network change will be beneficial at the first place, especially considering that it is not justified by theory. It might be only beneficial for chess and shogi, for example.

**Comment by justinwong098** [comment]
@isty2e 
Your quoted section says that only AlphaGo and AlphaGo Zero exploited symmetry. They said AlphaZero does not "augment training data"; if they only wanted this to apply to the Go version of AlphaZero, I'd expect they would say so. Also, see the Domain Knowledge section, where symmetry augmentation isn't mentioned. "AlphaZero did not use any form of domain knowledge beyond the points listed [in the Domain Knowledge section]."

I'm not convinced that dropping evaluation would create a large imbalance between training and self-play. Can you expand on this point further? Because as GCP has mentioned, evaluation is "a major source of latency." Why would dropping evaluation make the imbalance worse?

**Comment by isty2e** [comment]
@justinwong098 Ugh, you're right. I some how read it as "AlphaZero and AlphaGo Zero" instead of "AlphaGo and AlphaGo Zero".

Regarding the imbalance, I think major arguments are in the comment I wrote above: "If training is slower, the quality of network will not increase sufficiently fast, thereby resulting in self-plays of lower quality. On the other hand, if self-play generation is slower, then the network is likely to be overfitted, and considering that less training can be more desirable, the result can be questionable. It would be hard to figure out such details and balance them."

**Comment by billyswong** [comment]
What interested me is this paragraph
> Moves are selected in proportion to the root visit count. Dirichlet noise Dir(α) was added to the prior probabilities in the root node; this was scaled in inverse proportion to the approximate number of legal moves in a typical position, to a value of α = {0.3, 0.15, 0.03} for chess, shogi and Go respectively.

So the handcrafted first-30-moves-more-random rule is deleted/generalized.

**Comment by justinwong098** [comment]
@gcp 
>It's surprising to me this approach works. We've seen the playing strength drop greatly during training.

My (uneducated) speculation is that during self-play the bot learns certain moves/techniques which make it weaker overall (explaining the drop in strength), due to using the technique too much. However, once enough games have accumulated, it's able to generalize better and know when to use a given technique, allowing it to beat the current best network.
On the other hand, if a new network were to be used for self-play without evaluation, this still would also work, since during the self-play games, the network would also learn to use the moves/techniques (learned previously) sparingly.

**Comment by justinwong098** [comment]
@isty2e 
I'm still not understanding. If we drop evaluation, do you think training will become faster than self-play, or the reverse? And, as you see it, what is the state of things right now? Do you currently see it as balanced, or is one currently outweighing the other?

**Comment by gcp** [comment]
>So the handcrafted first-30-moves-more-random rule is deleted/generalized.

The Dirichlet noise was already there, so it's effectively deleted.

**Comment by isty2e** [comment]
@justinwong098 It's not a matter of dropping evaluation, rather a continuous and simultaneous training. Currently, clearly testing is bottlenecking, but once testing is distributed and training is automated then this will be settled.

**Comment by gcp** [comment]
To be honest the new paper is very sorely lacking in details.

**Comment by isty2e** [comment]
Yes, including the ANN architecture.

**Comment by bood** [comment]
@gcp Would it be reasonable to try that no-evaluation approach with the current data we have to see how it goes?

**Comment by gcp** [comment]
Probably. I need to think about the current situation a bit.

**Comment by isty2e** [comment]
Then it becomes ambiguous how much training steps we’ll take... This is effectively handled in AZ by balancing the speeds IMO, but without evaluation it is just unclear what to do for LZ.

**Comment by gcp** [comment]
I'm not sure what you mean. AZ doesn't balance the speeds at all. If they use more training HW, training goes faster. If they'd use more self-players, self-play would go faster. There's no specific indication they are specifically balanced out against each other, as the parts run asynchronously.

**Comment by isty2e** [comment]
@gcp I mean, they could tune the parameters and number of allocated hardwares. You can think of something like the number of effective training steps of a specific self-play game, and this will increase if less TPUs are used for self-play and increase if self-play becomes faster. And it seems like the number of training steps influence the strength of the resulting network. Thus, the ratio between computational resources allocated to NN training and self-play can be considered as a hyperparameter.

**Comment by gcp** [comment]
Yes, that's possible. They talk about tuning the hyper-parameters, so it's very possible if you'd ask them "why 800 or why 1600" the answer is just "that was stronger'.

**Comment by isty2e** [comment]
@gcp So my point is that the relative speed of training and self-play can be an important thing which is not published at all. And clearly it will be hard to optimize that value for LZ, if it is possible at all. Anyway, their no-evaluation strategy works because the effective training step is adequate for that process, but it will be hard to translate it into LZ. I think it's safer to replicate AGZ rather than AZ, and it is unclear if AZ is more effective for go than AGZ after all.

**Comment by gcp** [comment]
"Moves are selected in proportion to the root visit count."

Oh, so it's not deleted. They do it everywhere.

**Comment by marcocalignano** [comment]
Try the no evaluation approach was discarded few days ago when we had that discussion on letting the weaker network let generate self played games. Now everything is changed because someone else did it too? I think we should have a plan for it. Not change our strategy every time anyone write a paper about it. 
@gcp if your problem is testing overhead we are ready for a test on a larger scale (10 -15 people). with 2654 games per hour we get 400 games in ~9 mins so really little overhead.

**Comment by gcp** [comment]
>Now everything is changed because someone else did it too? I think we should have a plan for it. Not change our strategy every time anyone write a paper about it.

I am interested in what works. I discarded it earlier because there was no proof no testing was a good idea. Now there's some proof it can work.

But I find the argument from @isty2e compelling, i.e. that we have no tight locking between the two steps and it is not obvious that it would work for our setup.

**Comment by jkiliani** [comment]
The distributed testing setup will definitely be highly valuable for the project, no matter what is decided regarding the evaluation, since it gives a quick and easy solution to track progress. It would be a comparatively simple matter to use it for recalibrating the whole progress curve, especially those steps which are associated with high uncertainty because they were terminated early when a new network was significantly better.

One potential takeaway I see here would be to update instead of not updating in case of new networks of roughly the same strength, which seems to happen often. This would give the benefit of adding more variation to the training data, since the play style of same strength networks may be very different.

**Comment by marcocalignano** [comment]
But this was my argument in the previous discussion only because the score is worst does not means that the network cannot learn from the self played game from this network. But I guess we needed a proof to believe it ;)

**Comment by justinwong098** [comment]
If I understand correctly, the issue raised by @isty2e is getting the right ratio between number of steps taken in training versus number of self-play games generated. Could we not just copy the ratio used in the paper? 700,000 steps for 21 million games, with perhaps some scaling to account for the fact that symmetry was not used?

**Comment by gcp** [comment]
It works out to about 30 000 games per iteration, which closely matches the "games are copied in bunches of 25 000".

**Comment by evanroberts85** [comment]
I always thought the sudden cut of for proportional root visit at 30 moves was strange, a slow fade seemed best to me but the new paper just suggests keeping it on all game? 

I wonder if they made any changes to the MCTS to make it more efficient, that allowed them to reduce the number of playouts to 800? I can not see how, maybe their training of the policy part of the network is just better that make the search more efficient? In any case their switch to a lower number of playouts is probably causes by another change, rather than just being generally better.

I am not surprised by this paper, I think self learning still has many more ways to optimise itself, I said I thought within the next few years we should be able to train a pro level programme in just 10,000 games and I stand by that.
.

**Comment by evanroberts85** [comment]
“The learning rate was set to 0.2 for each game, and was dropped three times (to 0.02, 0.002 and 0.0002 respectively) during the course of training.”

Is this information new, or is it the same as AlphaGo?

**Comment by gcp** [comment]
It's much higher than the previous one. (All the while, people in this thread were clamoring that our problem was that it was too high).

**Comment by jkiliani** [comment]
I cannot find in the paper any information on the exact configuration of the neural network (i.e. residual with x filters and y number of blocks, and split policy/value head). I only found

> Unless otherwise specified, the training and search algorithm and parameters are identical to AlphaGo Zero (29).

This sounds very vague to me. What do you guys make of this?

**Comment by gcp** [comment]
I think you can infer the neural network size from the speed per move. The chess network was smaller than the go network, for example.

Aja said on computer-go they will follow up with a "real" paper.

**Comment by lp200** [comment]
> We applied the AlphaZero algorithm to chess, shogi, and also Go. Unless otherwise specified, the same algorithm settings, network architecture, and hyper-parameters were used for all three  games.


I think that the difference in speed between chess and go is due to board size.  19x19,8x8

**Comment by gcp** [comment]
Ah, good point.

**Comment by jkiliani** [comment]
@gcp: Could this also be relevant to Leela Zero? I found this on the Deepmind website: https://arxiv.org/pdf/1711.09846.pdf

If this is applicable here, it could solve a lot of your tuning issues.

**Comment by evanroberts85** [comment]
@jkiliani 
Just reading through the first two sections that paper is very interesting. I have suggested a few times that paramaters should be continuously and automatically tested to see which works best. It does require though quite a lot of extra coding.

**Comment by gcp** [comment]
>I have suggested a few times that paramaters should be continuously and automatically tested to see which works best. It does require though quite a lot of extra coding.

The coding is trivial But we don't have the hardware for this: you have to repeat the first part of the process multiple times!

**Comment by billyswong** [comment]
From the paper linked by jkilliani:
> Population based training starts like parallel search, randomly sampling hyperparameters and weight initialisations. However, each training run asynchronously evaluates its performance periodically. If a model in the population is under-performing, it will exploit the rest of the population by replacing itself with a better performing model, and it will explore new hyperparameters by modifying the better model’s hyperparameters, before training is continued. This process allows hyperparameters to be optimised online, and the computational resources to be focused on the hyperparameter and weight space that has most chance of producing good  results. The result is a hyperparameter tuning method  that while very simple, results in faster learning, lower computational resources, and often better solutions.

It this is true, then all parameters are "trained and tuned" like the core weight file, without any need to do any restart.

**Comment by evanroberts85** [comment]
I wonder why, when they already had this code, did AlphaZero use fixed paramaters?

**Comment by jkiliani** [comment]
> I wonder why, when they already had this code, did AlphaZero use fixed paramaters?

Good question, my best guess is that they already had tested hyper parameters extensively in developing Alphago Zero, so they knew these would work well for this particular problem. Or they had already done the parameter tuning for AlphaZero by the time this method was perfected. We don't know. I don't know either whether this would work in case of Leela Zero, but just brought it to the developers' attention in case it would.

**Comment by evanroberts85** [comment]
AlphaZero was not about optimisation, but creating a universal codebase to win at several different domains. It was not designed to beat AlphaGo, that was just an accident as a result of certain simplifications that proved faster. That is not to say all simplifications are faster, not at all!

If they really wanted to create a specific Go AI that would improve as fast as possible, I am sure they know how to do this much faster than AlphaZero or AlphaGo using advanced optimisations techniques they have learned, but that was and is not their aim, and they have nothing left to prove as they are already world champion at Go.

**Comment by tarriel** [comment]
Would it be possible to setting up some testing for a 5x5 board? There would be no guarantee it would behave the same as 19x19, but might make tractable testing possible for mortals like most of us.

**Comment by cgibbard** [comment]
I suspect that there is a regime early on where the chance of a decrease in strength is higher. When so many outcome-changing mistakes are being made that the result of the game can be thrown several times in a single game, the outcome is only just barely statistically meaningful. There were many games where the player who randomly played better for most of the game ended up poking out an eye of their own large group and losing, making the result overall misleading.

As the level of play improves, the more likely that each bad play will be punished, and that an advantage won't be thrown away, which then makes the likelihood that an individual game will be misleading become smaller and the the quality of training better.

So I suspect that testing for strength improvement becomes less necessary over time. We'll have to see. :)

**Comment by amj** [comment]
@gcp "Moves are selected in proportion to the root visit count."

That still leaves a lot of wiggle room for their temperature parameter, e.g. "in what proportion, exactly?"  What do you make of it?

**Comment by alreadydone** [comment]
@amj If the temperature is not 1 they say "exponentiated visit count" as in the AGZ paper.

**Comment by gcp** [comment]
@amj "That still leaves a lot of wiggle room for their temperature parameter, e.g. "in what proportion, exactly?" What do you make of it?"

It's either t=0 or t=1. t=0 means best move gets 100% probability, t=1 means it's a simple weighted pick.

**Comment by jkiliani** [comment]
You mean N^(1/t) I think. N^t would mean each move has equal probability for t->0, which would not make for a good engine.

**Comment by TimYuenior** [comment]
I was under the impression that the target labels would change according to the given temperature with the same formula of N^(1/t). I noticed in training.cpp that the target labels are always N/sum(N), so the case where t=1. Wouldn't the case of t=0 yield a vector with all zeros except for one entry that is one? I was curious if there is a reason why the target labels are always a simple weighted pick.
